### PR TITLE
fix(avro): correct parse `timestamp` data before Epoch

### PR DESCRIPTION
**Pull request by xuefengze**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

before:
insert timestamp before Epoch, will get error:
e.g. `1969-01-01 00:00:00`:
```
WARN   rw-streaming actor{otel.name="Actor 18" actor_id=18 prev_epoch=5850098781323264 curr_epoch=5850098793119744}:executor{otel.name="Source 1200002715 (actor 18)" actor_id=18}:source_parse_chunk{actor_id=18 source_id=1002}: risingwave_connector::parser: failed to parse non-pk column, padding with `NULL` error=Expected type `Some(Timestamptz)` but got `TimestampMillis(-31536000000)` for `` split_id="0" offset="2" column="c_timestamp" suppressed_count=0
```

Note:

While for the ancient `timestamp` data, it is incorrect:
`0001-01-01 00:00:00`(pg) ->  `-62135769600000`(kafka) ->`0000-12-30 00:00:00+00:00`(rw).

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/parser/unified/avro.rs [hunk 1]
```diff
@@ -181,19 +182,28 @@ impl<'a> AvroParseOptions<'a> {
...
+                    .into()
             }
 
             // ---- TimestampTz -----
             (Some(DataType::Timestamptz), Value::TimestampMillis(ms)) => {
-                i64_to_timestamptz(*ms).map_err(|_| create_error())?.into()
+                Timestamptz::from_millis(*ms)
+                    .ok_or(AccessError::Other(anyhow!(
+                        "timestamp with milliseconds {:?} * 1000 is out of range",
+                        ms
```
Comment (xiangjinwu) [inline src/connector/src/parser/unified/avro.rs:201]:
```suggestion
                        "timestamptz with milliseconds {ms} * 1000 is out of range",
```

File: src/connector/src/parser/unified/avro.rs [hunk 2]
```diff
@@ -181,19 +182,28 @@ impl<'a> AvroParseOptions<'a> {
...
+                    )))?
+                    .into()
             }
             (Some(DataType::Timestamptz), Value::TimestampMicros(us)) => {
```
Comment (xiangjinwu) [inline src/connector/src/parser/unified/avro.rs:205]: `Some(DataType::Timestamptz) | None` here and above.

**Review by xiangjinwu — APPROVED** [review]
Thanks for the fix!

**Review by tabVersion — APPROVED** [review]
lgtm

**Comment by xiangjinwu** [comment]
> While for the ancient `timestamp` data, it is incorrect:
> `0001-01-01 00:00:00`(pg) -> `-62135769600000`(kafka) ->`0000-12-30 00:00:00+00:00`(rw).

I tried this number in PostgreSQL 16:
```
test=# select extract(epoch from timestamptz '0001-01-01T00:00:00Z');
       extract       
---------------------
 -62135596800.000000
(1 row)

test=# select to_timestamp(-62135769600000/1e3);
       to_timestamp        
---------------------------
 0001-12-30 00:00:00+00 BC
(1 row)
```

So there are 2 issues:
* In the first step, `0001-01-01 00:00:00`(pg) shall be `-62135596800000` rather than `-62135769600000`.
* In the second step, it should be **displayed** as `0001 BC` rather than `0000`. They are essentially [the same year under different conventions](https://docs.rs/chrono/0.4.33/chrono/naive/struct.NaiveDate.html#:~:text=ISO%208601%20calendars%20has%20the%20year%200%2C%20which%20is%201%20BCE). If the fix is not easy we can consider it as a known incompatibility.

**Comment by xuefengze** [comment]
> In the first step, `0001-01-01 00:00:00`(pg) shall be `-62135596800000` rather than `-62135769600000`.

I think the issue is in `debezium-pg/confluent-pg` connector.

**Pull request merged**